lib/crypto-algorithms/megolm.js

```diff
@@ -0,0 +1,171 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+"use strict";
+
+/**
+ * Defines m.olm encryption/decryption
+ *
+ * @module crypto-algorithms/olm
+ */
+
+var q = require("q");
+
+var utils = require("../utils");
+var base = require("./base");
+
+var MEGOLM_ALGORITHM = "m.megolm.v1.aes-sha2";
+
+/**
+ * Megolm encryption implementation
+ *
+ * @constructor
+ * @extends {module:crypto-algorithms/base.EncryptionAlgorithm}
+ *
+ * @param {string} deviceId The identifier for this device.
+ * @param {module:crypto} crypto crypto core
+ * @param {module:OlmDevice} olmDevice olm.js wrapper
+ *
+ */
+function MegolmEncryption(deviceId, crypto, olmDevice) {
+    base.EncryptionAlgorithm.call(this, deviceId, crypto, olmDevice);
+    this._outboundSessionId = null;
+}
+utils.inherits(MegolmEncryption, base.EncryptionAlgorithm);
+
+/**
+ * @inheritdoc
+ * @param {string[]} roomMembers list of currently-joined users in the room
+ * @return {module:client.Promise} Promise which resolves when setup is complete
+ */
+MegolmEncryption.prototype.initRoomEncryption = function(roomMembers) {
+    // nothing to do here.
+    return q();
+};
+
+
+/**
+ * @private
+ * @param {string} roomId
+ */
+MegolmEncryption.prototype._ensureOutboundSession = function(roomId) {
+    if (this._outboundSessionId) {
+        return;
+    }
+    var session_id = this._olmDevice.createOutboundGroupSession();
+    this._outboundSessionId = session_id;
+
+    var key = this._olmDevice.getOutboundGroupSessionKey(session_id);
+
+    // TODO: initiate key-sharing
+
+    console.log(
+        'Created outbound session. Add with window.mxMatrixClientPeg.' +
+            'matrixClient._crypto._olmDevice.addInboundGroupSession("' +
+            [
+                roomId, this._olmDevice.deviceCurve25519Key, session_id,
+                key.key, key.chain_index
+            ].join('", "') +
+            '")'
+    );
+
+    this._olmDevice.addInboundGroupSession(
+        roomId, this._olmDevice.deviceCurve25519Key, session_id,
+        key.key, key.chain_index
+    );
+};
+
+/**
+ * @inheritdoc
+ *
+ * @param {module:models/room} room
+ * @param {string} eventType
+ * @param {object} plaintext event content
+ *
+ * @return {object} new event body
+ */
+MegolmEncryption.prototype.encryptMessage = function(room, eventType, content) {
+    this._ensureOutboundSession(room.roomId);
+
+    var payloadJson = {
+        room_id: room.roomId,
+        type: eventType,
+        content: content
+    };
+
+    var ciphertext = this._olmDevice.encryptGroupMessage(
+        this._outboundSessionId, JSON.stringify(payloadJson)
+    );
+
+    var encryptedContent = {
+        algorithm: MEGOLM_ALGORITHM,
+        sender_key: this._olmDevice.deviceCurve25519Key,
+        body: ciphertext,
+        session_id: this._outboundSessionId,
+        signature: "FIXME",
```

Comment:
Maybe add a comment with a TODO? probably not worth it though. I think it's fairly obvious what needs to be done to fix it.